File: src/ios/LDBCordova.m
#import "LDBCordova.h"

#import "LowlaDB/LowlaDB.h"

@interface LDBCordova ()
@property NSMutableDictionary *liveCursors;
@end

@implementation LDBCordova

- (void) pluginInitialize
{
    _liveCursors = [NSMutableDictionary dictionary];
    [LDBClient enableNotifications:YES];
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc addObserver:self selector:@selector(notifyLive:) name:LDBClientDidChangeCollectionNotification object:nil];
}

- (void) client_version:(CDVInvokedUrlCommand *)command
{
	LDBClient *client = [[LDBClient alloc] init];
	NSString *version = [client version];
	
	CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:version];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) lowla_load:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSString *json = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        if (json) {
            LDBClient *client = [[LDBClient alloc] init];
            [client loadJson:json];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: missing json"];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) db_collectionNames:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSString *dbName = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        if (dbName) {
            LDBClient *client = [[LDBClient alloc] init];
            LDBDb *db = [client getDatabase:dbName];
            NSArray *names = [db collectionNames];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:names];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: missing dbName"];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) db_dropDatabase:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSString *dbName = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        if (dbName) {
            LDBClient *client = [[LDBClient alloc] init];
            [client dropDatabase:dbName];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: missing dbName"];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) db_sync:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        NSString *server = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        if (server) {
            [LDBClient sync:server notify:^(LDBSyncStatus status, NSString *message) {
                CDVPluginResult *pluginResult;
                
                switch (status) {
                    case LDBSyncStatus_PUSH_STARTED:
                        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"pushBegin"];
                        [pluginResult setKeepCallbackAsBool:YES];
                        break;
                    case LDBSyncStatus_PUSH_ENDED:
                        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"pushEnd"];
                        [pluginResult setKeepCallbackAsBool:YES];
                        break;
                    case LDBSyncStatus_PULL_STARTED:
                        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"pullBegin"];
                        [pluginResult setKeepCallbackAsBool:YES];
                        break;
                    case LDBSyncStatus_PULL_ENDED:
                        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"pullEnd"];
                        [pluginResult setKeepCallbackAsBool:YES];
                        break;
                    case LDBSyncStatus_WARNING:
                        // Don't do anything yet
                        break;
                    case LDBSyncStatus_ERROR:
                        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:message];
                        break;
                    case LDBSyncStatus_OK:
                        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                        break;
                }
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }];
        }
        else {
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"No server specified"];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }
    }];
}

- (void) collection_findAndModify:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSString *dbName = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        NSString *collName = [command argumentAtIndex:1 withDefault:nil andClass:[NSString class]];
        NSDictionary *querySpec = [command argumentAtIndex:2 withDefault:nil andClass:[NSDictionary class]];
        NSDictionary *opSpec = [command argumentAtIndex:3 withDefault:nil andClass:[NSDictionary class]];
        
        if (dbName && collName && opSpec) {
            @try {
                LDBClient *client = [[LDBClient alloc] init];
                LDBDb *db = [client getDatabase:dbName];
                LDBCollection *coll = [db getCollection:collName];
                LDBObject *query = nil;
                if (querySpec) {
                    query = [LDBObject objectWithDictionary:querySpec];
                }
                LDBObject *ops = [LDBObject objectWithDictionary:opSpec];
                LDBWriteResult *wr = [coll update:query object:ops];
                if (0 < [wr documentCount]) {
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[[wr document:0] asJson]];
                }
                else {
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                }
            }
            @catch (NSException *e) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[e reason]];
            }
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: missing dbName or collName"];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) collection_insert:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSString *dbName = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        NSString *collName = [command argumentAtIndex:1 withDefault:nil andClass:[NSString class]];
        id data = [command argumentAtIndex:2];
        
        if (dbName && collName) {
            LDBClient *client = [[LDBClient alloc] init];
            LDBDb *db = [client getDatabase:dbName];
            LDBCollection *coll = [db getCollection:collName];
            LDBWriteResult *wr;
            @try {
                if ([data isKindOfClass:[NSDictionary class]]) {
                    LDBObject *object = [LDBObject objectWithDictionary:data];
                    wr = [coll insert:object];
                }
                else if ([data isKindOfClass:[NSArray class]]) {
                    NSMutableArray *arr = [NSMutableArray arrayWithCapacity:[(NSArray *)data count]];
                    [data enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
                        if ([obj isKindOfClass:[NSDictionary class]]) {
                            LDBObject *object = [LDBObject objectWithDictionary:obj];
                            [arr addObject:object];
                        }
                    }];
                    wr = [coll insertArray:arr];
                }
                NSMutableArray *docs = [NSMutableArray array];
                for (int i = 0 ; i < [wr documentCount] ; ++i) {
                    [docs addObject:[[wr document:i] asJson]];
                }
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:docs];
            }
            @catch (NSException *e) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[e reason]];
            }
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: missing dbName or collName"];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) collection_remove:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSString *dbName = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
        NSString *collName = [command argumentAtIndex:1 withDefault:nil andClass:[NSString class]];
        NSDictionary *query = [command argumentAtIndex:2 withDefault:nil andClass:[NSDictionary class]];
        
        if (dbName && collName) {
            LDBClient *client = [[LDBClient alloc] init];
            LDBDb *db = [client getDatabase:dbName];
            LDBCollection *coll = [db getCollection:collName];
            LDBObject *queryObj = nil;
            if (query) {
                queryObj = [LDBObject objectWithDictionary:query];
            }
            LDBWriteResult *wr = [coll remove:queryObj];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:[wr documentCount]];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: missing dbName or collName"];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

static LDBObject *convertSortSpec(id sortSpec) {
    if ([sortSpec isKindOfClass:[NSString class]]) {
        return [[[LDBObjectBuilder builder] appendInt:1 forField:sortSpec] finish];
    }
    else if ([sortSpec isKindOfClass:[NSArray class]]) {
        LDBObjectBuilder *builder = [LDBObjectBuilder builder];
        __block BOOL foundOne = NO;
        [(NSArray *)sortSpec enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            if ([obj isKindOfClass:[NSString class]]) {
                [builder appendInt:1 forField:obj];
                foundOne = YES;
            }
            else if ([obj isKindOfClass:[NSArray class]]) {
                NSArray *arr = (NSArray *)obj;
                if (1 <= [arr count]) {
                    id name = [arr objectAtIndex:0];
                    if ([name isKindOfClass:[NSString class]]) {
                        int asc = 1;
                        if (2 <= [arr count]) {
                            id order = [arr objectAtIndex:1];
                            if ([order isKindOfClass:[NSNumber class]]) {
                                asc = 0 < [(NSNumber *)order intValue] ? 1 : -1;
                            }
                        }
                        [builder appendInt:asc forField:name];
                        foundOne = YES;
                    }
                }
            }
        }];
        if (foundOne) {
            return [builder finish];
        }
        else {
            return nil;
        }
    }
    else {
        return nil;
    }
}

static LDBCursor *convertCursorSpec(NSDictionary *cursorSpec) {
    if (!cursorSpec) {
        return nil;
    }
    NSDictionary *coll = (NSDictionary *)[cursorSpec objectForKey:@"_collection"];
    NSString *dbName = (NSString *)[coll objectForKey:@"dbName"];
    NSString *collName = (NSString *)[coll objectForKey:@"collectionName"];
    if (dbName && collName) {
        LDBClient *client = [[LDBClient alloc] init];
        LDBDb *db = [client getDatabase:dbName];
        LDBCollection *coll = [db getCollection:collName];
        NSDictionary *filterDict = (NSDictionary *)[cursorSpec objectForKey:@"_filter"];
        LDBObject *filter = nil;
        if (filterDict) {
            filter = [LDBObject objectWithDictionary:filterDict];
        }
        LDBCursor *cursor = [[LDBCursor alloc] initWithCollection:coll query:filter keys:nil];
        NSDictionary *options = (NSDictionary *)[cursorSpec objectForKey:@"_options"];
        if (options) {
            id sortSpec = [options objectForKey:@"sort"];
            if (sortSpec) {
                LDBObject *sort = convertSortSpec(sortSpec);
                if (sort) {
                    cursor = [cursor sort:sort];
                }
            }
            id limitSpec = [options objectForKey:@"limit"];
            if (limitSpec && [limitSpec isKindOfClass:[NSNumber class]]) {
                int limit = [limitSpec intValue];
                if (0 < limit) {
                    cursor = [cursor limit:limit];
                }
            }
            id showPendingSpec = [options objectForKey:@"showPending"];
            if (showPendingSpec && [showPendingSpec isKindOfClass:[NSNumber class]]) {
                BOOL showPending = [showPendingSpec boolValue];
                if (showPending) {
                    cursor = [cursor showPending];
                }
            }
        }
        return cursor;
    }
    return nil;
}

- (void) cursor_count:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        
        NSDictionary *cursorSpec = [command argumentAtIndex:0 withDefault:nil andClass:[NSDictionary class]];
        LDBCursor *cursor = convertCursorSpec(cursorSpec);
        if (cursor) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:(int)[cursor count]];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: cursor has invalid structure"];
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) cursor_each:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult *pluginResult;
        NSDictionary *cursorSpec = [command argumentAtIndex:0 withDefault:nil andClass:[NSDictionary class]];
        LDBCursor *cursor = convertCursorSpec(cursorSpec);
        if (cursor) {
            while ([cursor hasNext]) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[[cursor next] asJson]];
                [pluginResult setKeepCallbackAsBool:YES];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:NO];
                [pluginResult setKeepCallbackAsBool:NO];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: cursor has invalid structure"];
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) addLiveCursor:(NSString *)ns withCallback:(NSString *)callbackId
{
    NSMutableArray *arr = [self.liveCursors objectForKey:ns];
    if (nil == arr) {
        arr = [NSMutableArray array];
        [self.liveCursors setObject:arr forKey:ns];
    }
    [arr addObject:callbackId];
}

- (void) notifyLive:(NSNotification *)notification
{
    NSString *ns = [[notification userInfo] objectForKey:@"ns"];
    
    NSMutableArray *arr = [self.liveCursors objectForKey:ns];
    if (nil != arr) {
        [arr enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"notify"];
            [pluginResult setKeepCallbackAsBool:YES];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:obj];
        }];
    }
}

- (void) cursor_on:(CDVInvokedUrlCommand *)command
{
    [self.commandDelegate runInBackground:^{
        NSDictionary *cursorSpec = [command argumentAtIndex:0 withDefault:nil andClass:[NSDictionary class]];
        NSDictionary *coll = (NSDictionary *)[cursorSpec objectForKey:@"_collection"];
        NSString *dbName = (NSString *)[coll objectForKey:@"dbName"];
        NSString *collName = (NSString *)[coll objectForKey:@"collectionName"];
        
        CDVPluginResult *pluginResult;
        if (dbName && collName) {
            [self addLiveCursor:[NSString stringWithFormat:@"%@.%@", dbName, collName] withCallback:command.callbackId];
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"started"];
            [pluginResult setKeepCallbackAsBool:YES];
        }
        else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Internal error: cursor has invalid structure"];
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}
@end

File: src/android/LDBCordova.java
package io.lowla.lowladb;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.lowla.lowladb.platform.android.Integration;

public class LDBCordova extends CordovaPlugin {
    private Map<String, Class<? extends CordovaOperation>> ops;
    private Map<String, List<CallbackContext>> liveCursors;

    static class MapAdapter implements Map<String, Object> {
        private JSONObject obj;

        MapAdapter(JSONObject obj) {
            this.obj = obj;
        }

        @Override
        public void clear() {
            throw new UnsupportedOperationException();
        }

        @Override
        public boolean containsKey(Object key) {
            return obj.has(key.toString());
        }

        @Override
        public boolean containsValue(Object value) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Set<Entry<String, Object>> entrySet() {
            throw new UnsupportedOperationException();
        }

        @Override
        public Object get(Object key) {
            Object answer = obj.opt(key.toString());
            if (answer instanceof JSONObject) {
                answer = new MapAdapter((JSONObject)answer);
            }
            return answer;
        }

        @Override
        public boolean isEmpty() {
            return obj.length() == 0;
        }

        @Override
        public Set<String> keySet() {
            Set<String> answer = new HashSet<String>();
            Iterator<String> it = obj.keys();
            while (it.hasNext()) {
                answer.add(it.next());
            }
            return answer;
        }

        @Override
        public void putAll(Map map) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Object remove(Object key) {
            throw new UnsupportedOperationException();
        }

        @Override
        public int size() {
            return obj.length();
        }

        @Override
        public Collection<Object> values() {
            throw new UnsupportedOperationException();
        }

        @Override
        public Object put(String key, Object value) {
            throw new UnsupportedOperationException();
        }
    }

    abstract class CordovaOperation implements Runnable {
        @Override
        public final void run() {
            try {
                runOperation();
            }
            catch (JSONException e) {
                callbackContext.error(e.getMessage());
            }
        }

        void setArgs(JSONArray args) {
            this.args = args;
        }

        void setCallbackContext(CallbackContext callbackContext) {
            this.callbackContext = callbackContext;
        }

        protected abstract void runOperation() throws JSONException;

        protected JSONArray args;
        protected CallbackContext callbackContext;
    }

    @Override
    protected void pluginInitialize() {
        Integration.INSTANCE.init(this.cordova.getActivity().getApplicationContext(), "LowlaDB");

        ops = new HashMap<String, Class<? extends CordovaOperation>>();
        ops.put("db_collectionNames", Db_CollectionNames.class);
        ops.put("db_dropDatabase", Db_DropDatabase.class);
        ops.put("db_sync", Db_Sync.class);
        ops.put("collection_findAndModify", Collection_FindAndModify.class);
        ops.put("collection_insert", Collection_Insert.class);
        ops.put("collection_remove", Collection_Remove.class);
        ops.put("cursor_count", Cursor_Count.class);
        ops.put("cursor_each", Cursor_Each.class);
        ops.put("cursor_on", Cursor_On.class);
        ops.put("lowla_load", Lowla_Load.class);

        liveCursors = new HashMap<String, List<CallbackContext>>();

        LDBClient.enableNotifications(true);
        LDBClient.addCollectionChangedListener(new LDBClient.CollectionChangedListener() {
            public void onCollectionChanged(String ns) {
                notifyLive(ns);
            }
        });
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        Class<? extends CordovaOperation> clazz = ops.get(action);
        if (null == clazz) {
            return false;
        }
        CordovaOperation op;
        try {
            Constructor<? extends CordovaOperation> ctor = clazz.getDeclaredConstructor(LDBCordova.class);
            op = ctor.newInstance(this);
        } catch (InstantiationException e) {
            return false;
        } catch (IllegalAccessException e) {
            return false;
        } catch (NoSuchMethodException e) {
            return false;
        } catch (InvocationTargetException e) {
            return false;
        }
        op.setArgs(args);
        op.setCallbackContext(callbackContext);
        cordova.getThreadPool().execute(op);
        return true;
    }

    class Db_CollectionNames extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String dbName = args.getString(0);
            LDBClient client = new LDBClient();
            LDBDatabase db = client.getDatabase(dbName);
            String[] names = db.collectionNames();
            callbackContext.success(new JSONArray(Arrays.asList(names)));
        }
    }

    class Db_DropDatabase extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String dbName = args.getString(0);
            LDBClient client = new LDBClient();
            client.dropDatabase(dbName);
            callbackContext.success();
        }
    }

    class Db_Sync extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String server = args.getString(0);
            if (null == server) {
                callbackContext.error("No server specified");
                return;
            }
            LDBClient.sync(server, new LDBClient.SyncNotifier() {
                @Override
                public void notify(LDBClient.LDBSyncStatus status, String message) {
                    PluginResult pluginResult = null;
                    switch (status) {
                        case PUSH_STARTED:
                            pluginResult = new PluginResult(PluginResult.Status.OK, "pushBegin");
                            pluginResult.setKeepCallback(true);
                            break;
                        case PUSH_ENDED:
                            pluginResult = new PluginResult(PluginResult.Status.OK, "pushEnd");
                            pluginResult.setKeepCallback(true);
                            break;
                        case PULL_STARTED:
                            pluginResult = new PluginResult(PluginResult.Status.OK, "pullBegin");
                            pluginResult.setKeepCallback(true);
                            break;
                        case PULL_ENDED:
                            pluginResult = new PluginResult(PluginResult.Status.OK, "pullEnd");
                            pluginResult.setKeepCallback(true);
                            break;
                        case WARNING:
                            // Nothing for now
                            break;
                        case ERROR:
                            pluginResult = new PluginResult(PluginResult.Status.ERROR, message);
                            break;
                        case OK:
                            pluginResult = new PluginResult(PluginResult.Status.OK, (String)null);
                            break;
                    }
                    if (null != pluginResult) {
                        callbackContext.sendPluginResult(pluginResult);
                    }
                }
            });
        }
    }

    class Collection_FindAndModify extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String dbName = args.getString(0);
            String collName = args.getString(1);
            JSONObject querySpec = args.optJSONObject(2);
            JSONObject opSpec = args.getJSONObject(3);
            try {
                LDBClient client = new LDBClient();
                LDBDatabase db = client.getDatabase(dbName);
                LDBCollection coll = db.getCollection(collName);
                LDBObject query = null;
                if (null != querySpec) {
                    query = LDBObject.objectWithMap(new MapAdapter(querySpec));
                }
                LDBObject ops = LDBObject.objectWithMap(new MapAdapter(opSpec));
                LDBWriteResult wr = coll.update(query, ops);
                if (0 < wr.getDocumentCount()) {
                    callbackContext.success(wr.getDocument(0).asJson());
                } else {
                    callbackContext.success();
                }
            }
            catch (Exception e) {
                callbackContext.error(e.getMessage());
            }
        }
    }

    class Collection_Insert extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String dbName = args.getString(0);
            String collName = args.getString(1);

            LDBClient client = new LDBClient();
            LDBDatabase db = client.getDatabase(dbName);
            LDBCollection coll = db.getCollection(collName);
            LDBWriteResult wr;
            try {
                JSONArray arrData = args.optJSONArray(2);
                JSONObject objData = args.optJSONObject(2);
                if (null != objData) {
                    LDBObject obj = LDBObject.objectWithMap(new MapAdapter(objData));
                    wr = coll.insert(obj);
                }
                else {
                    LDBObject[] arr = new LDBObject[arrData.length()];
                    for (int i = 0 ; i < arrData.length() ; ++i) {
                        arr[i] = LDBObject.objectWithMap(new MapAdapter(arrData.getJSONObject(i)));
                    }
                    wr = coll.insertArray(arr);
                }
                JSONArray answer = new JSONArray();
                for (int i = 0 ; i < wr.getDocumentCount() ; ++i) {
                    answer.put(wr.getDocument(i).asJson());
                }
                callbackContext.success(answer);
            }
            catch (Exception e) {
                callbackContext.error(e.getMessage());
            }
        }
    }

    class Collection_Remove extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String dbName = args.getString(0);
            String collName = args.getString(1);
            JSONObject querySpec = args.optJSONObject(2);

            LDBClient client = new LDBClient();
            LDBDatabase db = client.getDatabase(dbName);
            LDBCollection coll = db.getCollection(collName);
            LDBObject query = null;
            if (null != querySpec) {
                query = LDBObject.objectWithMap(new MapAdapter(querySpec));
            }
            LDBWriteResult wr = coll.remove(query);
            callbackContext.success(wr.getDocumentCount());
        }
    }

    private LDBObject convertSortSpec(Object sortSpec) throws JSONException {
        if (sortSpec instanceof String) {
            return new LDBObjectBuilder().appendInt((String)sortSpec, 1).finish();
        }
        else if (sortSpec instanceof JSONArray) {
            boolean foundOne = false;
            LDBObjectBuilder builder = new LDBObjectBuilder();
            for (int i = 0 ; i < ((JSONArray)sortSpec).length() ; ++i) {
                Object obj = ((JSONArray)sortSpec).get(i);
                if (obj instanceof String) {
                    builder.appendInt((String)obj, 1);
                    foundOne = true;
                }
                else if (obj instanceof JSONArray) {
                    JSONArray arr = (JSONArray)obj;
                    if (1 <= arr.length()) {
                        String name = arr.optString(0);
                        if (null != name) {
                            int asc = 1;
                            if (2 <= arr.length()) {
                                asc = 0 < arr.optInt(1, 1) ? 1 : -1;
                            }
                            builder.appendInt(name, asc);
                            foundOne = true;
                        }
                    }
                }
            }
            if (foundOne) {
                return builder.finish();
            }
        }
        return null;
    }

    private LDBCursor convertCursorSpec(JSONObject cursorSpec) throws JSONException {
        if (null == cursorSpec) {
            return null;
        }
        JSONObject collSpec = cursorSpec.getJSONObject("_collection");
        String dbName = collSpec.getString("dbName");
        String collName = collSpec.getString("collectionName");

        LDBClient client = new LDBClient();
        LDBDatabase db = client.getDatabase(dbName);
        LDBCollection coll = db.getCollection(collName);

        JSONObject filterSpec = cursorSpec.optJSONObject("_filter");
        LDBObject filter = null;
        if (null != filterSpec) {
            filter = LDBObject.objectWithMap(new MapAdapter(filterSpec));
        }
        LDBCursor cursor = coll.find(filter);

        JSONObject options = cursorSpec.optJSONObject("_options");
        if (null != options) {
            Object sortSpec = options.get("sort");
            if (null != sortSpec) {
                LDBObject sort = convertSortSpec(sortSpec);
                if (null != sort) {
                    cursor = cursor.sort(sort);
                }
            }
            int limit = options.optInt("limit");
            if (0 < limit) {
                cursor = cursor.limit(limit);
            }
            boolean showPending = options.optBoolean("showPending");
            if (showPending) {
                cursor = cursor.showPending();
            }
        }
        return cursor;
    }

    class Cursor_Count extends CordovaOperation {
        @Override
        public void runOperation() throws JSONException {
            JSONObject cursorSpec = args.getJSONObject(0);

            LDBCursor cursor = null;
            try {
                cursor = convertCursorSpec(cursorSpec);
                if (null != cursor) {
                    callbackContext.success((int) cursor.count());
                } else {
                    callbackContext.error("Cursor has invalid structure");
                }
            }
            finally {
                if (null != cursor) {
                    cursor.close();
                }
            }
        }
    }

    class Cursor_Each extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            JSONObject cursorSpec = args.getJSONObject(0);

            LDBCursor cursor = null;
            try {
                cursor = convertCursorSpec(cursorSpec);
                if (null != cursor) {
                    while (cursor.hasNext()) {
                        PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, cursor.next().asJson());
                        pluginResult.setKeepCallback(true);
                        callbackContext.sendPluginResult(pluginResult);
                    }
                    PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, false);
                    pluginResult.setKeepCallback(false);
                    callbackContext.sendPluginResult(pluginResult);
                } else {
                    callbackContext.error("Cursor has invalid structure");
                }
            }
            finally {
                if (null != cursor) {
                    cursor.close();
                }
            }
        }
    }

    private void addLiveCursor(String ns, CallbackContext callbackContext) {
        List<CallbackContext> callbacks = liveCursors.get(ns);
        if (null == callbacks) {
            callbacks = new ArrayList<CallbackContext>();
            liveCursors.put(ns, callbacks);
        }
        callbacks.add(callbackContext);
    }

    private void notifyLive(String ns) {
        List<CallbackContext> callbacks = liveCursors.get(ns);
        if (null != callbacks) {
            for (CallbackContext callbackContext : callbacks) {
                PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, "notify");
                pluginResult.setKeepCallback(true);
                callbackContext.sendPluginResult(pluginResult);
            }
        }
    }

    class Cursor_On extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            JSONObject cursorSpec = args.getJSONObject(0);
            JSONObject collSpec = cursorSpec.getJSONObject("_collection");
            String dbName = collSpec.getString("dbName");
            String collName = collSpec.getString("collectionName");

            addLiveCursor(dbName + "." + collName, callbackContext);
            PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, "started");
            pluginResult.setKeepCallback(true);
            callbackContext.sendPluginResult(pluginResult);
        }
    }

    class Lowla_Load extends CordovaOperation {
        @Override
        protected void runOperation() throws JSONException {
            String json = args.getString(0);
            LDBClient client = new LDBClient();
            client.loadJson(json);
            callbackContext.success();
        }
    }
}
